EDM-1254: Add user documentation for rollout policies (#989)
flightctl/flightctl: docs/user/rollout-policies.md

# Rollout Device Selection and Rollout Disruption Budget

## Overview

> When performing a rollout using `flightctl`, it is crucial to manage which devices participate in the rollout and how much disruption is acceptable. This document outlines the device selection process and the concept of the rollout disruption budget to ensure controlled and predictable rollouts.

Reviewer: "which devices participate in the rollout" makes it sound like not all devices in a fleet need to be rolled out.  In the end, all of them will be.

---

## Rollout Device Selection

### **Device Targeting**

> A rollout applies only to devices that belong to a fleet. Each device can belong to only a single fleet. Since rollout definitions are done at the fleet level, the selection process determines which devices within a fleet will participate in the rollout based on label criteria.

Reviewer: "which devices within a fleet will participate in the rollout" - same as above - all devices in a fleet will participate.

- **Labels**: Devices with specific metadata labels can be targeted for rollouts.
- **Fleet Membership**: Rollouts apply only to devices within the specified fleet.

### **Device Selection Strategy**

Currently, Flightctl supports only the **BatchSequence** strategy for device selection. This strategy defines a stepwise rollout process where devices are added in batches based on specific criteria.

Batches are executed sequentially. After each batch completes, execution proceeds to the next batch only if the success rate of the previous batch meets or exceeds the configured **success threshold**. The success rate is determined as:

```text
# of successful rollouts in the batch / # of devices in the batch >= success threshold
```

> A **successful rollout** is when a device completes the rollout process successfully.

Reviewer: This is obvious.  It would be more helpful to specify the criteria.  For example, the status field(s) that you look at.

Author: Will remove

### **Limit in Device Selection**

> Each batch in the **BatchSequence** strategy uses a `limit` parameter to define how many devices should be included in the batch. The limit can be specified in two ways:

Reviewer 2: `limit` is optional, since the batch can be defined entirely based on the label selector.
As explained here, it would seem like it's a required field.

Author: OK

- **Absolute number**: A fixed number of devices to be selected.

> - **Percentage**: A percentage of the total device population that matches the label criteria.

Reviewer 2: Since as it's explained  just below, "Percentage" can be used without the selector, the base definition can be contradictory.

```suggestion
- **Percentage**: Percentage of the total matching device population to be selected.
```

Author: OK

  - If a `selector` with labels is provided, the percentage is calculated based on the number of devices that match the label criteria within the fleet.
  - If no `selector` is provided, the percentage is applied to all devices in the fleet.

### **Success Threshold**

> The **success threshold** defines the percentage of successfully updated devices required to continue the rollout. If the success rate falls below this threshold, the rollout may be paused or rolled back to prevent further failures.

Reviewer: What do you mean by "rolled back"?  We don't have rollback.

Author: Right

> ### **Example YAML Configuration (Fleet Spec)**

Reviewer: It would be helpful also to describe the example in words.

```yaml
apiVersion: v1alpha1
kind: Fleet
metadata:
  name: default
spec:
  selector:
    matchLabels:
      fleet: default
  rolloutPolicy:
    deviceSelection:
      strategy: 'BatchSequence'
      sequence:
        - selector:
            matchLabels:
              site: madrid
          limit: 1  # Absolute number
        - selector:
            matchLabels:
              site: madrid
          limit: 80%  # Percentage of devices matching the label criteria within the fleet
        - limit: 50%  # Percentage of all devices in the fleet
        - selector:
            matchLabels:
              site: paris
        - limit: 80%
        - limit: 100%
    successThreshold: 95%
```

---

## Rollout Disruption Budget

### **Definition**

A rollout disruption budget defines the acceptable level of service impact during a rollout. This ensures that a deployment does not take down too many devices at once, maintaining overall system stability.

### **Disruption Budget Parameters**

> 1. **Group By**: Defines how devices are grouped when applying the disruption budget.

Reviewer 2: We should indicate that the "groupBy" entries should match label keys within the fleet devices.

An maybe provide a detailed example out of the YAML definition below (eg. define what groups exist, and that the budget refers to each of them independently).

Author: OK

> # Rollout Device Selection and Rollout Disruption Budget

Reviewer 2: The reference in `user/README.md` to the definition of the rollout policies needs to be updated.

Currently it links to `managing-fleets.md#defining-rollout-policies` which does not exist, and in this PR it should point to `rollout-policies.md`
